lib/logstash/codecs/telemetry.rb

```diff
@@ -450,23 +615,152 @@ def decode(data)
+                    # Stringify operation copes with nested hashes too.
+                    # .stringify in rails is what I am looking for, but this
+                    # is not rails.
+                    #
+                    ev = msg_out.clone
+                    ev[:end_time] = msg_out[:end_time]
+                    ev[:content] = row_out
+                    ev[:type] = row_decoder_name[1]
+                    ev[:path] = table[:policy_path]
+                    yield LogStash::Event.new(JSON.parse(ev.to_json))
+
+                  end # End of iteration over rows
+
+                rescue Exception => e
+                  @logger.warn("Failed to decode telemetry row",
+                               :policy_path => table[:policy_path],
+                               :decoder => row_decoder_name,
+                               :exception => e, :stacktrace => e.backtrace)
+                end # End of exception handling of row decode
+
+                @logger.debug? && @logger.debug("Iteration end")
+
+              else # No decoder is available
+
+                @logger.debug? &&
+                  @logger.debug("No decoder available",
+                                :policy_path => table[:policy_path])
+
+              end # End of cases where a decoder is available, or not
+
+            end # End of iteration over each table
+
+          rescue Exception => e
+            @logger.warn("Failed to decode telemetry header",
+                         :data => decompressed_unit,
+                         :exception => e, :stacktrace => e.backtrace)
+          end
+
+        when TSCodecTLVTypeV2::TS_CODEC_TLV_TYPE_V2_GPB_KV
+          #####!!!!decode_kv(msg)
+          v, msg = msg.unpack("a#{l}a*")
+
+          if @data_compressed == 1
+            decompressed_unit = @zstream.inflate(v)
+            @logger.debug? &&
+            @logger.debug("Parsed message", :zmsgtype => t, :zmsglength => l,
+                          :msglength => decompressed_unit.length,
+                          :decompressor => @zstream)
           else
-            # default case, something's gone awry
-            @logger.error("Resetting connection on unknown TLV type",
-                           :type => t)
-            raise 'Unexpected message type in TLV:. Reset connection'
+            decompressed_unit = v
           end
 
-        end # end look through TLVs
+          msg_gpb = Telemetry.new
+          evs = Array.new
+
+          begin
+            msg_out = msg_gpb.parse(decompressed_unit).to_hash
+            tables = msg_out.delete(:tables)
+            tables.each do |table|
+
+              @logger.debug? &&
+                @logger.debug("Message policy paths",
+                              :collection_id => msg_out[:collection_id],
+                              :base_path => msg_out[:base_path],
+                              :msg_timestamp => msg_out[:msg_timestamp])
+
+              decode_kv_table(table,evs)
```

Comment:
As per above, you will want to pass down the collection start time here, such that it, or a more specific timestamp on the way down will be included in the event.